substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,231 @@
+        for (PremainMethod premainMethod : premainMethods) {
+            try {
+                Object[] args = premainMethod.args;
+                if (premainOptions.containsKey(premainMethod.className)) {
+                    args[0] = premainOptions.get(premainMethod.className);
+                }
+                // premain method must be static
+                premainMethod.method.invoke(null, args);
+            } catch (Throwable t) {
+                VMError.shouldNotReachHere("Fail to execute " + premainMethod.className + ".premain", t);
+            }
+        }
+    }
+
+    /**
+     * This class is a dummy implementation of {@link Instrumentation} interface. It serves as the
+     * registered premain method's second parameter. At native image runtime, no actual
+     * instrumentation work can do. So all the methods here are empty.
+     */
+    public static class NativeImageNoOpRuntimeInstrumentation implements Instrumentation {
+
+        @Override
+        public void addTransformer(ClassFileTransformer transformer, boolean canRetransform) {
+        }
+
+        @Override
+        public void addTransformer(ClassFileTransformer transformer) {
+        }
+
+        @Override
+        public boolean removeTransformer(ClassFileTransformer transformer) {
+            return false;
+        }
+
+        @Override
+        public boolean isRetransformClassesSupported() {
+            return false;
+        }
+
+        @Override
+        public void retransformClasses(Class<?>... classes) throws UnmodifiableClassException {
+        }
+
+        @Override
+        public boolean isRedefineClassesSupported() {
+            return false;
+        }
+
+        @Override
+        public void redefineClasses(ClassDefinition... definitions) throws ClassNotFoundException, UnmodifiableClassException {
+        }
+
+        @Override
+        public boolean isModifiableClass(Class<?> theClass) {
+            return false;
+        }
+
+        @Override
+        public Class<?>[] getAllLoadedClasses() {
+            return new Class<?>[0];
+        }
+
+        @Override
+        public Class<?>[] getInitiatedClasses(ClassLoader loader) {
+            return new Class<?>[0];
+        }
+
+        @Override
+        public long getObjectSize(Object objectToSize) {
+            return 0;
+        }
+
+        @Override
+        public void appendToBootstrapClassLoaderSearch(JarFile jarfile) {
+        }
+
+        @Override
+        public void appendToSystemClassLoaderSearch(JarFile jarfile) {
+        }
+
+        @Override
+        public boolean isNativeMethodPrefixSupported() {
+            return false;
+        }
+
+        @Override
+        public void setNativeMethodPrefix(ClassFileTransformer transformer, String prefix) {
+        }
+
+        @Override
+        public void redefineModule(Module module, Set<Module> extraReads, Map<String, Set<Module>> extraExports, Map<String, Set<Module>> extraOpens, Set<Class<?>> extraUses,
```

Comment:
I feel this should be a hard failure that will explain to the users what they should do (preferable). 

Now, the question is how many agents will call this even when transformation is not allowed. I suspect not many.

Reply:
This method is called to declare extra exports/opens/uses/reads packages from module.
It is not rarely used. The agent may need extra access to certain modules.
We can call APIs from [ModuleNative](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/ModuleNative.java) to make checks only.

Reply:
I'd suggest `throw new UnsupportedOperationException()` based on the javadoc for the method https://docs.oracle.com/javase/8/docs/api/java/lang/instrument/Instrumentation.html#redefineClasses-java.lang.instrument.ClassDefinition...-

> [UnsupportedOperationException](https://docs.oracle.com/javase/8/docs/api/java/lang/UnsupportedOperationException.html) - if the current configuration of the JVM does not allow redefinition ([isRedefineClassesSupported()](https://docs.oracle.com/javase/8/docs/api/java/lang/instrument/Instrumentation.html#isRedefineClassesSupported--) is false) or the redefinition attempted to make unsupported changes